### PR TITLE
Cleanup in flat lighting

### LINKED ISSUE #7040: Dark and miscoloured nodeboxes
##### Issue type

 - Bug report
 
##### Minetest version

af89a25ff3527ccf0740926fe0b797c4403840b0

##### OS / Hardware

Operating system: ubuntu 16.04 LTS
CPU: Intel Core i5-4670T CPU @ 2.30GHz × 4

GPU model: Intel Haswell Desktop (integrated)
OpenGL version: 3.0

screenshot_20180215_102716

^ No smooth lighting

screenshot_20180215_102743

^ Smooth lighting, seems ok at first but the snow nodeboxes are actually darker and bluer away from the biome edge as is more obvious from higher up:

screenshot_20180215_102802

All effects, filters and shaders are off.
My guess is the recent changes to ambient occlusion @numberZero 
Setting 'sunlight_propagates = true' in the nodedef does not fix this.

### PR DESCRIPTION
**Pull request by numberZero**
Fixes flat lighting part of #7040
screenshot_20180218_004553

////////////////////
EDIT by paramat:
More explanation for reviewers from https://github.com/minetest/minetest/issues/7040#issuecomment-366447191
> nodeboxes use light one level brighter for drawing than their true light value. But to do that, undiminish_light is used, which clamps at LIGHT_MAX, which is 14. So solid nodes get 15, and nodeboxes are drawn at 14 under direct sunlight.

So the difference in light levels only became visible now that light levels 14 and 15 are visibly different due to #6696

### CODE REVIEW

**Comment by numberZero** [comment]
@paramat

File: src/mapblock_mesh.cpp [hunk 1]
```diff
@@ -130,18 +130,8 @@ static u8 getInteriorLight(enum LightBank bank, MapNode n, s32 increment,
...
-		light = undiminish_light(light);
-		--increment;
-	}
-	while(increment < 0)
-	{
-		light = diminish_light(light);
-		++increment;
-	}
-
+	if (light)
```
Comment (red-001) [inline src/mapblock_mesh.cpp]: imo it would be cleaner to write this as `light != 0` as we are using it as a number not a boolean.

Comment (numberZero) [reply src/mapblock_mesh.cpp]: Or `light > 0`; it’s unsigned so that’s all the same.

**Comment by Fixer-007** [comment]
Does not fix for me when smooth lighting is enabled it seems.

**Comment by numberZero** [comment]
> Also, one note on smooth lighting: it’s just the way it works. It can’t know that a face is exposed to direct sunlight, such logic would be complicated. I suppose that was imagined on adding the “non-trivial” label to the original issue.

**Comment by numberZero** [comment]
So yes, it doesn’t fix that.

**Comment by paramat** [comment]
Thanks will test.

**Comment by paramat** [comment]
The bug with smooth lighting under direct sunlight is very ugly so i think we will have to revert https://github.com/minetest/minetest/commit/2d9f0d344e45282638f9c1c1b57019c5e7f3999c#diff-1bf86c5e08b07bc255ad8698c565bab6 #6696 instead.
Thanks for this but 👎

**Comment by numberZero** [comment]
@paramat Don’t hurry, this PR is a cleanup, and only then a bugfix, so should probably be merged in any case.

**Comment by paramat** [comment]
Ok so you mean merge this and revert #6696 ?
Please could you explain in the first comment what is being done here and why? What is removed and changed seems important.
Are 'diminish_light' and 'undimish_light' used elsewhere other than for nodeboxes?
'getInterorLight' may be used for several things already, are they all checked?
If #6696 is reverted then why is this PR needed?

**Comment by paramat** [comment]
From IRC http://irc.minetest.net/minetest-dev/2018-02-20#i_5234309

12:21 numzero   paramat: (on #7051) `diminish_light` and `undiminish_light` were used in several places a long time ago; but currently, they are only used in `getInteriorLight`. And the way they are used is such that they can be replaced with just a few lines of code.
12:21 ShadowBot https://github.com/minetest/minetest/issues/7051 -- Cleanup in flat lighting by numberZero
12:22 numzero   Direct replacement would be clamping `light` to `LIGHT_MAX`, then, if it is non-zero, adding `increment` and then clamping to `LIGHT_MAX` again.
12:22 numzero   That would work *exactly* the same as the current code
12:24 numzero   My PR clamps to `LIGHT_SUN` instead, to fix #7040 for flat lighting. (The first clamping disappears as the original light level is never greater than `LIGHT_SUN`)
12:24 ShadowBot https://github.com/minetest/minetest/issues/7040 -- Dark and miscoloured nodeboxes
12:28 numzero   In the case #6696 is reverted, LIGHT_SUN and LIGHT_MAX are eqivalent again. So the only difference will be that light=LIGHT_SUN, increment=-1 will be treated as LIGHT_MAX (full brightness) instead of LIGHT_MAX-1
12:28 ShadowBot https://github.com/minetest/minetest/issues/6696 -- Update light decoding table size by numberZero
12:28 numzero   by so far, increment=1 is only ever used for selection box...
12:29 numzero   Nevertheless, I have yet another idea how to do things in a better way; working on that right now.

**Comment by Fixer-007** [comment]
7061 and this PR seems to fix the problems for me :+1:

**Comment by numberZero** [comment]
@Fixer-007 This should fix dark slabs (snow, mainly) that you mentioned in #7040 and not introduce any new bugs.

**Comment by Fixer-007** [comment]
Snow slabs without smooth lighting?

**Comment by paramat** [comment]
@Fixer-007 see https://github.com/minetest/minetest/issues/7040#issuecomment-366442722 onwards and numberZero's reply, that weirdness you see has always been there and is unavoidable with nodeboxes as explained.

> Maybe this is because nodeboxes get a single light level from inside and uses that for all faces, whereas cubic nodes have each face getting it's light level from the node neightbour?

> I'm beginning to wonder if it's always been like this and we didn't notice due to: 1. no smooth lighting on nodeboxes so they looked different anyway. 2. Appearence in direct sunlight was ok due to light levels 14 and 15 being the same. 3. When nodeboxes were given smooth lighting the smooth lighting disguised the difference.

**Comment by paramat** [comment]
Code looks good, will now test.

**Comment by paramat** [comment]
:+1: Fixes brightness mismatch in direct sunlight with smooth lighting disabled.

**Pull request merged**